examples/whl.py

```diff
@@ -0,0 +1,120 @@
+            distinfo = DistInfo.find(project_name, project_version, entry_names)
+            yield cls(name, distinfo, zf)
+
+    @contextlib.contextmanager
+    def _open_adjacent_tmp_for_write(self, path, **kwargs):
+        # type: (pathlib.Path, Any) -> Iterator[IO[Any]]
+        path.parent.mkdir(parents=True, exist_ok=True)
+        temp = path.with_name("{}.tmp.{}".format(path.name, self._name))
+        with temp.open(**kwargs) as f:
+            yield f
+        temp.replace(path)
+
+    def _open_target_for_write(self, path, binary=False):
+        # type: (pathlib.Path, bool) -> ContextManager[IO[Any]]
+        if binary:
+            kwargs = {"mode": "wb"}
+        else:
+            kwargs = {"mode": "w", "encoding": "utf-8"}
+        return self._open_adjacent_tmp_for_write(path, **kwargs)
+
+    def _open_csv_for_write(self, path):
+        # type: (pathlib.Path) -> ContextManager[IO[Any]]
+        if sys.version_info < (3,):
+            kwargs = {"mode": "wb"}
+        else:
+            kwargs = {"mode": "w", "newline": "", "encoding": "utf-8"}
+        return self._open_adjacent_tmp_for_write(path, **kwargs)
+
+    def _install_record_item(self, item, directory):
+        # type: (RecordItem, pathlib.Path) -> None
+        with self._zf.open(str(item.path)) as f:
+            data = f.read()
+        item.raise_for_validation(data)
+        target = directory.joinpath(item.path)
+        with self._open_target_for_write(target, binary=True) as f:
+            f.write(data)
+
+    def _iter_installed_record_items(self, directory):
+        # type: (pathlib.Path) -> Iterator[RecordItem]
+        reader = codecs.getreader("utf-8")
+        with self._zf.open(str(self._distinfo.record)) as f:
+            for item in parse_record_file(reader(f)):
+                self._install_record_item(item, directory)
+                yield item
+
+    def _iter_installed_scripts(self, directory):
+        # type: (pathlib.Path) -> Iterator[RecordItem]
+        return iter(())  # TODO: Implement me.
+
+    def _write_additional_metadata(self, directory):
+        # type: (pathlib.Path) -> Iterator[RecordItem]
+        installer = directory.joinpath(self._distinfo.installer)
+        with self._open_target_for_write(installer) as f:
+            f.write(self._name)
+        yield RecordItem(self._distinfo.installer, None, None)
+        # TODO: Write direct_url.json.
+
+    def _write_record(self, directory, installed_items):
+        # type: (pathlib.Path, Dict[pathlib.PurePosixPath, RecordItem]) -> None
+        record = self._distinfo.record
+        installed_items[record] = RecordItem(record, None, None)
+        with self._open_csv_for_write(directory.joinpath(record)) as f:
+            write_record_file(f, installed_items.values())
+
+    def install(self, directory):
+        # type: (pathlib.Path) -> None
+        items = {r.path: r for r in self._iter_installed_record_items(directory)}
```

Comment:
Do we need the iter keyword, feels like redundant to me 🤔

Reply:
I do like to spell out an iterator to distinguish it from sequences. The call site needs to be aware of their different lifetime characteristics, otherwise it’s too easy for a maintainer to accidentally iterate through an iterator twice.